### PR TITLE
Add IsFileLocal lightup, use it in `CA1708`

### LINKED ISSUE #6514: File-Scoped Types With The Same Name Trigger CA1708
### Analyzer

**Diagnostic ID**: [CA1708](https://learn.microsoft.com/en-us/dotnet/fundamentals/code-analysis/quality-rules/ca1708): `Identifiers should differ by more than case`

### Analyzer source

**SDK**: .NET 7

**Version**: 7.0.3

### Describe the bug

Basically what I say in the title. File-scoped types with the same name are triggering this issue.

### Steps To Reproduce

Create a console application, targeting .NET 7, and ensure the .csproj looks like this:

```xml
<Project Sdk="Microsoft.NET.Sdk">
	<PropertyGroup>
		<AnalysisMode>all</AnalysisMode>
		<TreatWarningsAsErrors>true</TreatWarningsAsErrors>
		<OutputType>Exe</OutputType>
		<TargetFramework>net7.0</TargetFramework>
		<ImplicitUsings>enable</ImplicitUsings>
		<Nullable>enable</Nullable>
	</PropertyGroup>
</Project>
```

Add two .cs files to the project, call them whatever you want (Class1.cs, Class2.cs), and put the following code into the files:

Class1.cs
```csharp
namespace ConsoleApp1;

file static class GeneratedName
{
    public const string Name = "Method";
}

public static class MethodName
{
    public static string RetrieveName() => GeneratedName.Name;
}
```

Class2.cs
```csharp
namespace ConsoleApp1;

file static class GeneratedName
{
    public const string Name = "Type";
}

public static class TypeName
{
    public static string RetrieveName() => GeneratedName.Name;
}
```

### Expected behavior

No issues should be created.

### Actual behavior

CA1708 is created on a build, something like this:

```
Severity	Code	Description	Project	File	Line	Suppression State
Error	CA1708	Names of 'Types' and 'ConsoleApp1.GeneratedName, ConsoleApp1.GeneratedName' should differ by more than case	ConsoleApp1	C:\Users\jason\source\repos\ConsoleApp1\ConsoleApp1\ConsoleApp1.csproj	1	Active
```

### PR DESCRIPTION
**Pull request by RenderMichael**
Fixes #6514

### CODE REVIEW

File: src/Utilities/Compiler/Lightup/IsFileLocalWrapper.cs [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+#if HAS_IOPERATION
+
+namespace Analyzer.Utilities.Lightup
+{
+    using System;
+    using System.Diagnostics.CodeAnalysis;
+    using Microsoft.CodeAnalysis;
+
+    [SuppressMessage("Performance", "CA1815:Override equals and operator equals on value types", Justification = "Not a comparable instance.")]
+    internal readonly struct IsFileLocalWrapper
```
Comment (sharwell) [inline src/Utilities/Compiler/Lightup/IsFileLocalWrapper.cs:12]: 💡 This should be a static class with an extension method named `IsFileLocal`:

```cs
public static bool IsFileLocal(this INamedTypeSymbol symbol)
  => s_isFileLocal(symbol);
```

Comment (RenderMichael) [reply src/Utilities/Compiler/Lightup/IsFileLocalWrapper.cs]: I added it to the existing `INamedTypeSymbolExtensions` class, I can add create a new one in the Lightup namespace if that's more desirable

File: src/NetAnalyzers/Core/Microsoft.CodeQuality.Analyzers/ApiDesignGuidelines/IdentifiersShouldDifferByMoreThanCase.cs [hunk 2]
```diff
@@ -96,8 +97,9 @@ private static void CheckNamespaceMembers(IEnumerable<INamespaceSymbol> namespac
             {
                 // Get all the potentially externally visible types in the namespace
                 IEnumerable<INamedTypeSymbol> typeMembers = @namespace.GetTypeMembers().Where(item =>
-                    Equals(item.ContainingAssembly, context.Compilation.Assembly) &&
-                    MatchesConfiguredVisibility(item, context.Options, context.Compilation));
+                Equals(item.ContainingAssembly, context.Compilation.Assembly) &&
+                MatchesConfiguredVisibility(item, context.Options, context.Compilation) &&
+                !item.IsFileLocal());
```
Comment (sharwell) [inline src/NetAnalyzers/Core/Microsoft.CodeQuality.Analyzers/ApiDesignGuidelines/IdentifiersShouldDifferByMoreThanCase.cs:102]:
```suggestion
                    Equals(item.ContainingAssembly, context.Compilation.Assembly) &&
                    MatchesConfiguredVisibility(item, context.Options, context.Compilation) &&
                    !item.IsFileLocal());
```

File: src/Utilities/Compiler/Extensions/INamedTypeSymbolExtensions.cs [hunk 3]
```diff
@@ -282,5 +283,9 @@ public static bool IsBenchmarkOrXUnitTestAttribute(this INamedTypeSymbol attribu
         public static bool IsTopLevelStatementsEntryPointType([NotNullWhen(true)] this INamedTypeSymbol? typeSymbol)
             => typeSymbol is not null &&
                typeSymbol.GetMembers().OfType<IMethodSymbol>().Any(m => m.IsTopLevelStatementsEntryPointMethod());
+
+        private static readonly Func<INamedTypeSymbol, bool> s_isFileLocal = LightupHelpers.CreateSymbolPropertyAccessor<INamedTypeSymbol, bool>(typeof(INamedTypeSymbol), nameof(IsFileLocal), fallbackResult: false);
```
Comment (sharwell) [inline src/Utilities/Compiler/Extensions/INamedTypeSymbolExtensions.cs:287]: 💡 We normally place fields at the top of the class

**Review by sharwell — APPROVED** [review]

File: src/Utilities/Compiler/Lightup/IsFileLocalWrapper.cs (hunk 1 above)
Comment (sharwell) [reply src/Utilities/Compiler/Lightup/IsFileLocalWrapper.cs]: This was a fine approach 👍

**Comment by RenderMichael** [comment]
Just realized I never added that `[WorkItem]` attribute, is it important @sharwell ?

**Review by sharwell — APPROVED** [review]

**Pull request merged**